napari/utils/chunk/_config.py

```diff
@@ -0,0 +1,137 @@
+"""AsyncConfig to configure asynchronous loading and the ChunkLoader.
+"""
+import errno
+import json
+import logging
+import os
+from pathlib import Path
+
+LOGGER = logging.getLogger("ChunkLoader")
```

Comment:
The intent of these log message is they never appear in any Napari-wide log. But they are something you can toggle onto get a dedicated log file with chunk async/ChunkLoader message, from whatever files are relevant. Basically a single unified trace of what's going on with async.

You toggle it on from the async config file like this:
```
    "log_path": "/Users/pbw/dev/napari/chunk.log",
```

So I'm not sure how to best achieve that. The idea of an optional single-purpose log file that's separate from main logging.

I also want to switch it to json format. Because the long messages are log and complicated like:
```
ChunkLoader._done: layer_id=140697234842720 data_id=140697506324208 data_level=7 indices=((0, 172, 1), (0, 157, 1))
```

I see some packages that do JSON logging with a custom formatter inside the native logging, but I haven't tried one.

Reply:
Ah, if your intent was to create a file-only logger that never logs to std out, then you should probably remove the `ChunkLoader` logger from the default [streamingHandler](https://docs.python.org/3.8/library/logging.handlers.html#streamhandler).  The point here is that we're using the global python logger... _any_ package can also be using it, and anybody using napari might just be using it as one of many tools that they are using.  So while your `log_path` works fine for your needs, you're adding a lot of extra info to the log, but you've also made it difficult for someone else to silence your stuff if they wanted to.  For instance, if I run `napari -vv` from this PR and do some stuff...

<details>

<summary>here's what I see ... </summary>

```
❯ python -m napari -vv
09:15:35 napari_ndtiffs.affine INFO Using pyopencl for affine transforms
09:15:35 napari_plugin_engine.manager INFO loaded 6 plugins:
  <module 'napari.plugins._builtins' from '/Users/talley/Dropbox (HMS)/Python/forks/napari/napari/plugins/_builtins.py'>
  <module 'napari_dv' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/napari_dv.py'>
  <module 'napari_s3zarr' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/napari_s3zarr/__init__.py'>
  <module 'ome_zarr' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/ome_zarr.py'>
  <module 'napari_svg' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/napari_svg/__init__.py'>
  <module 'napari_ndtiffs' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/napari_ndtiffs/__init__.py'>
  <module 'napari_plugin_engine' from '/Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/napari_plugin_engine/__init__.py'>
09:15:35 ChunkLoader INFO ImageSlice.__init__
09:15:35 ChunkLoader INFO ImageSlice.load_chunk: layer_id=5311660624 data_id=5311560480 data_level=0 indices=(0, (None, None, None), (None, None, None))
09:15:35 ChunkLoader INFO ChunkLoader.load_chunk
09:15:35 ChunkLoader INFO ImageSlice.chunk_loaded: accept layer_id=5311660624 data_id=5311560480 data_level=0 indices=(0, (None, None, None), (None, None, None))
09:15:40 ChunkLoader INFO ImageSlice.load_chunk: layer_id=5311660624 data_id=5311560480 data_level=0 indices=(12, (None, None, None), (None, None, None))
09:15:40 ChunkLoader INFO ChunkLoader.load_chunk
09:15:40 ChunkLoader INFO ImageSlice.chunk_loaded: accept layer_id=5311660624 data_id=5311560480 data_level=0 indices=(12, (None, None, None), (None, None, None))
09:15:41 ChunkLoader INFO ImageSlice.load_chunk: layer_id=5311660624 data_id=5311560480 data_level=0 indices=(27, (None, None, None), (None, None, None))
09:15:41 ChunkLoader INFO ChunkLoader.load_chunk
09:15:41 ChunkLoader INFO ImageSlice.chunk_loaded: accept layer_id=5311660624 data_id=5311560480 data_level=0 indices=(27, (None, None, None), (None, None, None))
09:15:41 ChunkLoader INFO ImageSlice.load_chunk: layer_id=5311660624 data_id=5311560480 data_level=0 indices=(45, (None, None, None), (None, None, None))
09:15:41 ChunkLoader INFO ChunkLoader.load_chunk
09:15:41 ChunkLoader INFO ImageSlice.chunk_loaded: accept layer_id=5311660624 data_id=5311560480 data_level=0 indices=(45, (None, None, None), (None, None, None))
09:15:53 PIL.Image DEBUG Importing BlpImagePlugin
09:15:53 PIL.Image DEBUG Importing BmpImagePlugin
09:15:53 PIL.Image DEBUG Importing BufrStubImagePlugin
09:15:53 PIL.Image DEBUG Importing CurImagePlugin
09:15:53 PIL.Image DEBUG Importing DcxImagePlugin
09:15:53 PIL.Image DEBUG Importing DdsImagePlugin
09:15:53 PIL.Image DEBUG Importing EpsImagePlugin
09:15:53 PIL.Image DEBUG Importing FitsStubImagePlugin
09:15:53 PIL.Image DEBUG Importing FliImagePlugin
09:15:53 PIL.Image DEBUG Importing FpxImagePlugin
09:15:53 PIL.Image DEBUG Image: failed to import FpxImagePlugin: No module named 'olefile'
09:15:53 PIL.Image DEBUG Importing FtexImagePlugin
09:15:53 PIL.Image DEBUG Importing GbrImagePlugin
09:15:53 PIL.Image DEBUG Importing GifImagePlugin
09:15:53 PIL.Image DEBUG Importing GribStubImagePlugin
09:15:53 PIL.Image DEBUG Importing Hdf5StubImagePlugin
09:15:53 PIL.Image DEBUG Importing IcnsImagePlugin
09:15:53 PIL.Image DEBUG Importing IcoImagePlugin
09:15:53 PIL.Image DEBUG Importing ImImagePlugin
09:15:53 PIL.Image DEBUG Importing ImtImagePlugin
09:15:53 PIL.Image DEBUG Importing IptcImagePlugin
09:15:53 PIL.Image DEBUG Importing JpegImagePlugin
09:15:53 PIL.Image DEBUG Importing Jpeg2KImagePlugin
09:15:53 PIL.Image DEBUG Importing McIdasImagePlugin
09:15:53 PIL.Image DEBUG Importing MicImagePlugin
09:15:53 PIL.Image DEBUG Image: failed to import MicImagePlugin: No module named 'olefile'
09:15:53 PIL.Image DEBUG Importing MpegImagePlugin
09:15:53 PIL.Image DEBUG Importing MpoImagePlugin
09:15:53 PIL.Image DEBUG Importing MspImagePlugin
09:15:53 PIL.Image DEBUG Importing PalmImagePlugin
09:15:53 PIL.Image DEBUG Importing PcdImagePlugin
09:15:53 PIL.Image DEBUG Importing PcxImagePlugin
09:15:53 PIL.Image DEBUG Importing PdfImagePlugin
09:15:53 PIL.Image DEBUG Importing PixarImagePlugin
09:15:53 PIL.Image DEBUG Importing PngImagePlugin
09:15:53 PIL.Image DEBUG Importing PpmImagePlugin
09:15:53 PIL.Image DEBUG Importing PsdImagePlugin
09:15:53 PIL.Image DEBUG Importing SgiImagePlugin
09:15:53 PIL.Image DEBUG Importing SpiderImagePlugin
09:15:53 PIL.Image DEBUG Importing SunImagePlugin
09:15:53 PIL.Image DEBUG Importing TgaImagePlugin
09:15:53 PIL.Image DEBUG Importing TiffImagePlugin
09:15:53 PIL.Image DEBUG Importing WebPImagePlugin
09:15:53 PIL.Image DEBUG Importing WmfImagePlugin
09:15:53 PIL.Image DEBUG Importing XbmImagePlugin
09:15:53 PIL.Image DEBUG Importing XpmImagePlugin
09:15:53 PIL.Image DEBUG Importing XVThumbImagePlugin
09:15:53 napari.plugins.io ERROR PluginError: Error in plugin 'builtins', hook 'napari_get_reader'
  Cause was: ValueError('Could not load "" \nReason: "cannot find loader for this HDF5 file"\nPlease see documentation at: http://pillow.readthedocs.io/en/latest/installation.html#external-libraries')
    in file: /Users/talley/miniconda3/envs/napdev/lib/python3.8/site-packages/imageio/plugins/pillow.py
    at line: 680
     author: napari team
      email: napari-steering-council@googlegroups.com
    package: napari
        url: https://napari.org
    version: 0.3.7.dev8+g9963d6bf

09:15:53 napari.plugins.io WARNING No plugin found capable of reading '/Users/talley/Desktop/gt_image_001_annotations.h5'.
09:15:53 napari.plugins.io ERROR (1) error occurred in plugins: 'builtins'. See full error logs in "Plugins → Plugin Errors..."
```
</details>

note how `ChunkLoader` is the only thing in there that is not prefixed by the originating package...


> Although I guess we can allow people to turn it on in the system-wide file, but it would likely swamp whatever else was going on.

Yeah, logs are often swampy anyway once you get enough libraries involved... so it is absolutely expected that the person paying attention to the log is going to want to create their own basicConfig to suit their needs.  But we're not the only library using the logging system, that's why I think it's important to namespace all loggers, as a "courtesy" to those trying to silence/listen to everything from a given package, for instance `napari.`

... regardless of what you end up doing with your special JSON file logger, I'd really prefer that you namespace this logger.

Reply:
Yeah using a better namespace makes sense. But I'm wondering what that namespace should be since it seems wrong for Image or ImageSlice to get the logger napari.utils.chunk.ChunkLoader doesn't it?

Maybe they should use a command like utils.chunk_loader.get_async_logger() and they don't care what the namespace is. Longer term they might want to emit their own "normal" log messages distinct from these async ones.

I might be able to get rid of other people logging to the ChunkLoader log file. It's been useful for development, but maybe it should go away eventually.

I was tempted up front to just not using logging at all. Since this was so specialized. Like if we switch to JSON format, emitting JSON message among non-JSON ones is probably not useful. So should it even use logging?

Reply:
> But I'm wondering what that namespace should be since it seems wrong for Image or ImageSlice to get the logger napari.utils.chunk.ChunkLoader?

that's basically what you've already done by naming it `ChunkLoader`, just without a prefix. If you want to have a global napari logger dedicated to chunk loading related tasks, then I think `utils.chunk_loader.get_chunk_logger()` is a great idea.  You can name it whatever you want, as long as it starts with `napari.`... 

> Maybe they should use a command like utils.chunk_loader.get_logger() and they don't care what the namespace is.

 My concerns here are not so much about people wanting to pay attention to the chunk loading logger, it's for people who _don't_ care about it... or in fact, don't care about _any_ of napari's loggers.  (this is also related to my request to have you decrease the logging level to debug for various things)

> Like if we switch to JSON format, emitting JSON message among non-JSON ones is probably not useful. So should it even use the logging module?

I think using logging is very reasonable here, and you might find some tips for JSON logging in https://docs.python.org/3/howto/logging-cookbook.html#implementing-structured-logging.

It's just that you need to take measures to make sure you either don't pollute the global streamhandler logger at all, or use standard namespacing conventions for those who'd like to silence it.

Reply:
I think the ideal behavior is it's totally off by default in the main log file, even if you use debug level. But if you are totally insane you can toggle it on somehow. So not sure if that's possible.

Does napari have a config file where users can configure logging? That seems like the important step. Where they don't have to edit code but they can pick and choose what to log.

I can set it to debug and to `napari.async` or `napari.chunk_loader` to start. But I think most people will want it off unless very specifically debugging async. And then the'll probably use the dedicated log file.

Reply:
> I think the ideal behavior is it's totally off by default in the main log file

What “main log file”?  We don’t have a main log file... I recognize that you’ve created a purpose specific file handler here in this PR, which is fine.  The default logging behavior of python is to log to stdout.

> Does napari have a config file where users can configure logging?

No, we don’t have any persistent storage of any kind.  So people interacting with the logging system now are all going to be using python’s builtin logging module, which is why it’s important that we adhere to conventions.

Reply:
By "main log file" I mean the main log output of napari, whether it's outputting to the stdout or a file or wherever. I mean "main" as opposed to the "async log file" which you toggle on with the log_file setting of the NAPARI_ASYNC=async.json config file.

Since there is a separate preferred way to get the async log, I think we should just remove it entirely from the main logging. I think 99% likely if someone is getting this output in their main log they are going to be totally annoyed by it. It's too verbose and spammy.

Early on you said remove it "from the default streamingHandler"? How/where do I do that? I think that's what we want to do here.

I'll also change the namespace to be napari.something, but I think permanently removing it is the best thing. Until/unless someone is asking for it to be included. Which I pretty much guarantee will be never

Reply:
I guess the way I would say it is merging this PR should not change the log output that napari produces. Maybe someday we will unify them. But for now I think the async/ChunkLoader log output should only be visible in the async log. Not sure how to arrange that though?